sycl/doc/extensions/experimental/sycl_ext_oneapi_bindless_images/ZE_experimental_bindless_images.asciidoc

```diff
@@ -0,0 +1,621 @@
+# Bindless Images
+
+:source-highlighter: coderay
+:coderay-linenums-mode: table
+:dpcpp: pass:[DPC++]
+
+// This section needs to be after the document title.
+:doctype: book
+:toc2:
+:toc: left
+:encoding: utf-8
+:lang: en
+
+:blank: pass:[ +]
+// Set the default source code type in this document to C++,
+// for syntax highlighting purposes.  This is needed because
+// docbook uses c++ and html5 uses cpp.
+:language: {basebackend@docbook:c++:cpp}
+
+
+- [Bindless Images](#bindless-images)
+  - [Notice](#notice)
+  - [Dependencies](#dependencies)
+  - [Status](#status)
+  - [Terminology](#terminology)
+  - [Overview](#overview)
+    - [Background](#background)
+    - [Specification](#specification)
+  - [Definitions](#definitions)
+    - [Interfaces](#interfaces)
+    - [Enums](#enums)
+  - [Programming example](#programming-example)
+    - [Non-USM](#non-usm)
+    - [USM](#usm)
+  - [Known Issues and Limitations](#known-issues-and-limitations)
+  - [Revision History](#revision-history)
+
+== Notice
+
+[%hardbreaks]
+Copyright (C) Codeplay. All rights reserved.
+
+Khronos(R) is a registered trademark and SYCL(TM) and SPIR(TM) are trademarks
+of The Khronos Group Inc.  OpenCL(TM) is a trademark of Apple Inc. used by
+permission by Khronos.
+
+
+== Dependencies
+
+This extension is written against the Level Zero v1.5.8 specification.
+
+== Status
+
+This is an experimental extension specification, intended to provide early
+access to features and gather community feedback.
+
+== Terminology
+
+For the purposes of this document, a bindless image is one which provides
+stateless access to the underlying data via image reference handles. At the
+application level, this allows the user to implement programs where the number
+of images is not known at compile-time, and store all handles to images --
+irrespective of varying formats and layouts -- in some container, e.g. a dynamic
+array.
+
+== Overview
+
+In this document, we propose the following changes to the Level Zero API:
+
+- The creation of images to be split into explicit allocation of image memory
+  and the creation of image handles from the previously allocated memory
+- Support for creation of images on linearly allocated memory backed by USM
+- Addition of image memory copy functions that allow copying to and from image
+  memory before an image handle's construction
+
+Currently, in Level Zero, the `zeImageCreate` performs the image memory
+allocation and image handle generation. Moreover, this function only allows for
+the allocation of image memory in an implementation-specific layout. Delivering
+the features proposed in this document will improve the functionality and
+flexibility of the Level Zero API.
+
+
+=== Background
+
+The DPC++ bindless images extension has sought to provide the flexibility of
```

Comment:
Does the term "bindless image" refer to the possibility of passing variable number of images to kernel ? Or to the allocation flow - how backing memory is allocated  and usd for image ? 

It looks to me the extension defined in this document describes the new way of allocating images - and new memory handles. It does not need to be coupled with bindless images ( allowing  passing variable count of image handles to kernels at runtime ) - that are orthogonal to the way of how images are allocated.

Reply:
> Does the term "bindless image" refer to the possibility of passing variable number of images to kernel ?

Yes.

> Or to the allocation flow - how backing memory is allocated and usd for image ?

No. However, the allocation flow that we've proposed is necessary for our ability to implement the bindless images extension in the DPC++ Level Zero backend.

> It does not need to be coupled with bindless images

The new way of allocating images, and hence the new handles, is necessary for device buffers containing device image handles. We were unaware until yesterday that the existing `ze_image_handle_t` is a complex type and thus not valid on the device. With this, the extension will also require for the user to have access to device image handles, e.g. through
```
ze_result_t zeImageGetDeviceHandle(
      ze_image_handle_t image,            /// [in]
      unsigned long long* device_handle   /// [out] – type TBD
);
```